Short-circuit the rendering of fully transparent symbol tiles.

AdamSzofranMSFT opened the pull request:
When there are large clusters of colliding symbolic icons and labels, many of them are culled by being set fully transparent but their geometry is still rendered. In some cases, entire tiles full of symbols can be made transparent but are still rendered. This change looks for those fully transparent symbol tiles and avoids rendering them at all which can greatly reduce GL context commands (including draws). For example, for Bing Maps in downtown Seattle at LOD 17 the total number of rendering commands can be reduced by 20-25% with this change.

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to [related issues](https://github.com/maplibre/maplibre-gl-js/issues/96).
 - [ ] (n/a) Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] (n/a) Write tests for all new functionality.
 - [ ] (n/a) Document any changes to public APIs.
 - [x] Post benchmark scores. 
 ![benchmark result](https://raw.githubusercontent.com/AdamSzofranMSFT/maplibre-gl-js/users/AdamSzofranMSFT/PR-data/users/AdamSzofranMSFT/LayerSymbolWithSortKey-benchmark-1376b12.png)
 - [x] Manually test the debug page.
 - [ ] Add an entry to `CHANGELOG.md` under the `## main` section.

HarelM commented:
How would this change affect query render features?

AdamSzofranMSFT commented:
> How would this change affect query render features?
That sounds like a great question... about which I have no clue! So, I'll see if I can find an answer. Thanks.

AdamSzofranMSFT commented:
> > How would this change affect query render features?

> @HarelM is your question about the impact of this PR on the `test-query` tests and the two query related benchmarks? Or something else? Thanks.

FWIW, the `test-query` tests all pass and the two query related benchmarks don't show any significant difference.

HarelM commented:
The tests may pass if there isn't a test there to simulate this "extreme" condition.
The question here, due to the change in rendering etc, if this somehow affect the logic related to querying these hidden-now-removed features.
I'm not sure I know how to test this though...

AdamSzofranMSFT commented:
@HarelM are you referring to the [`queryRenderedFeatures`](https://maplibre.org/maplibre-gl-js-docs/api/map/#map#queryrenderedfeatures) API specifically? Thanks.

HarelM commented:
Yes, exactly this API.

AdamSzofranMSFT commented:
This PR waits until nearly the last moment to avoid rendering the fully transparent symbol tiles. It doesn't keep them from being loaded or stored or considered by render queries as far as I can tell.

Up to the point of rendering in [`drawLayerSymbols`](https://github.com/maplibre/maplibre-gl-js/pull/1757/files#diff-55d56e80d4928a3c2d33beab5b184f578149627933338bce9460a6389b986e4b) the only thing this PR does is add some additional state to the symbol tiles to help track the aggregate transparency of the symbols in the tile.

It looks like there might be one possible place where side-effects could come into play. The early out from `drawLayerSymbols` happens before it calls `updateLineLabels`. If that call affects the results of `queryRenderedFeatures` then perhaps I can move the early out after `updateLineLabels` to preserve the side-effect.

AdamSzofranMSFT commented on src/render/draw_symbol.ts:
```diff
@@ -291,7 +291,7 @@ function drawLayerSymbols(
         if (!bucket) continue;
         const buffers = isText ? bucket.text : bucket.icon;
-        if (!buffers || !buffers.segments.get().length) continue;
+        if (!buffers || !buffers.segments.get().length || !buffers.hasVisibleVertices) continue;
```
Would it be better to move the `hasVisibleVertices` early out to after the `updateLineLabels` call around line 345? That might preserve any side-effects caused by `updateLineLabels` if they are necessary to guarantee correct results from the `queryRenderedFeatures` API.

HarelM replied:
IDK, you'll need to investigate I guess...

AdamSzofranMSFT replied:
Was not necessary to move.

HarelM commented:
I would also like to see tests that are making sure this behavior is as expected - both the old behavior and the new one.

AdamSzofranMSFT commented:
Thanks. Will try to author some unit tests to determine the impact of this PR on `queryRenderedFeatures`, if any.

AdamSzofranMSFT commented:
FWIW, I instrumented the `LayerSymbolWithSortKey` benchmark to call `queryRenderedFeatures` on every interation and it gets the same result with or without the changes in this PR. I know hacking the benchmark isn't enough; I need to write a unit test that shows the same. But at least it shows that probably no harm has been done.

AdamSzofranMSFT commented:
@HarelM, I added a new map unit test that confirms that the result of `queryRenderedFeatures()` is unaffected by the render optimization in this PR. The test compares the number of features returned by `queryRenderedFeatures()` with and without the optimization enabled and it gets the same result both times.

HarelM commented on src/render/draw_symbol.ts:
```diff
@@ -251,6 +251,18 @@ function getSymbolProgramName(isSDF: boolean, isText: boolean, bucket: SymbolBuc
+
+// to allow mocking, the function above is embedded in a mockable object
+// reference: https://medium.com/@qjli/how-to-mock-specific-module-function-in-jest-715e39a391f4
+export const mockableFunctions = {
```
This is not needed in production code, mother is the above method.
In jest and javascript you can mock anything anywhere.

AdamSzofranMSFT replied:
I couldn't get the mock to work without both of these, but I'll try again.

HarelM replied:
You are testing from "too far". If your are writing a test against map class you shouldn't be mocking anything that low, you should be able to test the functionality according to the relevant use case. Testing the method you changed by testing the file you changed should allow you to be able to mock what you need (i.e. the bucket)

AdamSzofranMSFT replied:
In order to confirm that my rendering optimization doesn't impact the query API, I need both the rendering pipeline and query API to be fully functional. Is that why you recommended adding a query integration test instead of adding a map test? Should I just move the test over there?

HarelM replied:
If you tested the performance change locally I think it might be good enough, besides the performance, there's a need to test the new functionality, to make sure in the future that it won't break if someone changes the code...

AdamSzofranMSFT replied:
Resolved by most recent commits.

HarelM commented:
Thanks for taking the time to review my comments!
I would focus on writing a unit test around the method you changed and render test to check the functionally.

HarelM commented:
Also there are also the query tests which can be used here as well I guess.

AdamSzofranMSFT commented:
@HarelM, I removed the test I added to map.test.ts and recreated it (in sprit, at least) as a query test. I also added a simple test to draw_symbol.test.ts to confirm that no rendering occurs for symbol buffers without the `hasVisibleVertices` flag set.

HarelM commented on src/symbol/placement.ts:
```diff
@@ -946,8 +946,14 @@ export class Placement {
-        if (bucket.hasIconData()) bucket.icon.opacityVertexArray.clear();
+        if (bucket.hasTextData()) {
+            bucket.text.opacityVertexArray.clear();
+            bucket.text.hasVisibleVertices = false;
```
Do we need this extra field? Can we check if the array is empty?

AdamSzofranMSFT replied:
Checking.

HarelM commented:
Last comment before I merge...

AdamSzofranMSFT commented:
> Last comment before I merge...

All good now @HarelM ?

HarelM commented:
Yes, Thanks!
I saw your initial comment about checking the flag but Github doesn't notify when you edit a comment, so I didn't have the last amswer.
Thanks for the ping!
